server/src-lib/Hasura/Events/Lib.hs

```diff
@@ -166,49 +168,86 @@ retryAfterHeader = "Retry-After"
 
 initEventEngineCtx :: Int -> DiffTime -> STM EventEngineCtx
 initEventEngineCtx maxT fetchI = do
-  q <- TQ.newTQueue
-  c <- newTVar 0
-  return $ EventEngineCtx q c maxT fetchI
-
-forkEventQueueProcessors
+  c <- newTVar maxT
+  return $ EventEngineCtx c fetchI
+
+-- | Service events from our in-DB queue.
+--
+-- There are a few competing concerns and constraints here; we want to...
+--   - fetch events in batches for lower DB pressure
+--   - don't fetch more than N at a time (since that can mean: space leak, less
+--     effective scale out, possible double sends for events we've checked out 
+--     on exit (TODO clean shutdown procedure))
+--   - try not to cause webhook workers to stall waiting on DB fetch
+--   - limit webhook HTTP concurrency per HASURA_GRAPHQL_EVENTS_HTTP_POOL_SIZE
+processEventQueue
   :: (HasVersion) => L.Logger L.Hasura -> LogEnvHeaders -> HTTP.Manager-> Q.PGPool
   -> IO SchemaCache -> EventEngineCtx 
-  -> IO (Immortal.Thread, Immortal.Thread)
-  -- ^ returns: (pushEvents handle, consumeEvents handle)
-forkEventQueueProcessors logger logenv httpMgr pool getSchemaCache eectx = do
-  (,) <$> forkImmortal "pushEvents" logger pushEvents
-      <*> forkImmortal "consumeEvents" logger consumeEvents
+  -> IO void
+processEventQueue logger logenv httpMgr pool getSchemaCache EventEngineCtx{..} = do
+  events0 <- popEventsBatch
+  go events0 0 False
   where
-    -- FIXME proper backpressure. See: #3839 
-    pushEvents = forever $ do
-      let EventEngineCtx q _ _ fetchI = eectx
-      eventsOrError <- runExceptT $ Q.runTx pool (Q.RepeatableRead, Just Q.ReadWrite) fetchEvents
-      case eventsOrError of
-        Left err     -> L.unLogger logger $ EventInternalErr err
-        Right events -> atomically $ mapM_ (TQ.writeTQueue q) events
-      sleep fetchI
-
-    -- TODO this has all events race. How do we know this is correct? Document.
-    consumeEvents = forever $
-      -- ensure async exceptions from link only raised between iterations of forever block:
-      mask_ $ do
-        event <- atomically $ do
-          let EventEngineCtx q _ _ _ = eectx
-          TQ.readTQueue q
-        -- FIXME proper backpressure. See: #3839 
-        t <- async $ runReaderT (processEvent event) (logger, httpMgr, eectx)
-        -- Make sure any stray exceptions are at least logged via 'forkImmortal':
-        link t
-
-    -- NOTE: Blocks in tryWebhook if >= _eeCtxMaxEventThreads invocations active.
+    fetchBatchSize = 100
+    popEventsBatch = do
+      let run = runExceptT . Q.runTx pool (Q.RepeatableRead, Just Q.ReadWrite)
+      run (fetchEvents fetchBatchSize) >>= \case
+          Left err -> do 
+            L.unLogger logger $ EventInternalErr err
+            return []
+          Right events -> 
+            return events
+
+    go :: [Event] -> Int -> Bool -> IO void
```

Comment:
@jberryman I am slighly confused how are we capping the memory/threads in flight. IIUC, we keep fetching events and get blocked on processing them because of `HASURA_GRAPHQL_EVENTS_HTTP_POOL_SIZE`

Reply:
How is this more beneficial than using the in-memory queue that we had before `_eeCtxEventQueue`

Reply:
The main issue with the old version is we fetch from DB at intervals without regard for how fast we're processing. If webhooks are very slow we will check out more and more events to process, maybe leaking all memory.

Here we prefetch at most two batches, we don't recurse on `go` and fetch another batch until we've (concurrently) processed the last batched (or to be precise, when there is some extra capacity in our workers).

I agree this could be more clear! will improve comments , then would be great if you could take another look

Reply:
> Here we prefetch at most two batches, we don't recurse on go and fetch another batch until we've (concurrently) processed the last batched (or to be precise, when there is some extra capacity in our workers).

Is it because it gets blocked here https://github.com/hasura/graphql-engine/pull/4013/files#diff-a5793e7f6a1e13c4a19d13cae27dae88R215?

Reply:
yep we fork a new worker only when capacity is freed, I'll beef out comments